src/topology/metric_space/infsep.lean

```diff
@@ -0,0 +1,464 @@
+end
+
+theorem einfsep_eq_top_iff : s.einfsep = ∞ ↔ s.subsingleton :=
+⟨subsingleton_of_einfsep_eq_top, subsingleton.einfsep⟩
+
+theorem nontrivial.einfsep_ne_top (hs : s.nontrivial) : s.einfsep ≠ ∞ :=
+by { contrapose! hs, rw not_nontrivial_iff, exact subsingleton_of_einfsep_eq_top hs }
+
+theorem nontrivial.einfsep_lt_top (hs : s.nontrivial) : s.einfsep < ∞ :=
+by { rw lt_top_iff_ne_top, exact hs.einfsep_ne_top }
+
+theorem einfsep_lt_top_iff : s.einfsep < ∞ ↔ s.nontrivial :=
+⟨nontrivial_of_einfsep_lt_top, nontrivial.einfsep_lt_top⟩
+
+theorem einfsep_ne_top_iff : s.einfsep ≠ ∞ ↔ s.nontrivial :=
+⟨nontrivial_of_einfsep_ne_top, nontrivial.einfsep_ne_top⟩
+
+lemma le_einfsep_of_forall_dist_le {d} (h : ∀ (x y ∈ s) (hxy : x ≠ y), d ≤ dist x y) :
+   ennreal.of_real d ≤ s.einfsep :=
+le_einfsep $
+λ x hx y hy hxy, (edist_dist x y).symm ▸ ennreal.of_real_le_of_real (h x hx y hy hxy)
+
+end pseudo_metric_space
+
+section emetric_space
+variables [emetric_space α] {x y z : α} {s t : set α} {C : ℝ≥0∞} {sC : set ℝ≥0∞}
+
+lemma einfsep_pos_of_finite [finite s] : 0 < s.einfsep :=
+begin
+  casesI nonempty_fintype s,
+  by_cases hs : s.nontrivial,
+  { rcases hs.einfsep_exists_of_finite with ⟨x, hx, y, hy, hxy, hxy'⟩,
+    exact hxy'.symm ▸ edist_pos.2 hxy },
+  { rw not_nontrivial_iff at hs,
+    exact hs.einfsep.symm ▸ with_top.zero_lt_top }
+end
+
+lemma relatively_discrete_of_finite [finite s] :
+  ∃ C (hC : 0 < C), ∀ (x y ∈ s) (hxy : x ≠ y), C ≤ edist x y :=
+by { rw ← einfsep_pos, exact einfsep_pos_of_finite }
+
+lemma finite.einfsep_pos (hs : s.finite) : 0 < s.einfsep :=
+by { letI := hs.fintype, exact einfsep_pos_of_finite }
+
+lemma finite.relatively_discrete (hs : s.finite) :
+  ∃ C (hC : 0 < C), ∀ (x y ∈ s) (hxy : x ≠ y), C ≤ edist x y :=
+by { letI := hs.fintype, exact relatively_discrete_of_finite }
+
+end emetric_space
+
+end einfsep
+
+section infsep
+open_locale ennreal
+open set function
+
+/-- The "infimum separation" of a set with an edist function. -/
+noncomputable def infsep [has_edist α] (s : set α) : ℝ := ennreal.to_real (s.einfsep)
+
+section has_edist
+variables [has_edist α] {x y : α} {s : set α}
+
+lemma infsep_zero : s.infsep = 0 ↔ s.einfsep = 0 ∨ s.einfsep = ∞ :=
+by rw [infsep, ennreal.to_real_eq_zero_iff]
+
+lemma infsep_nonneg : 0 ≤ s.infsep := ennreal.to_real_nonneg
+
+lemma infsep_pos : 0 < s.infsep ↔ 0 < s.einfsep ∧ s.einfsep < ∞ :=
+by simp_rw [infsep, ennreal.to_real_pos_iff]
+
+lemma subsingleton.infsep_zero (hs : s.subsingleton) : s.infsep = 0 :=
+by { rw [infsep_zero, hs.einfsep], right, refl }
+
+lemma nontrivial_of_infsep_pos (hs : 0 < s.infsep) : s.nontrivial :=
+by { contrapose hs, rw not_nontrivial_iff at hs, exact hs.infsep_zero ▸ lt_irrefl _ }
+
+lemma infsep_empty : (∅ : set α).infsep = 0 :=
+subsingleton_empty.infsep_zero
+
+lemma infsep_singleton : ({x} : set α).infsep = 0 :=
+subsingleton_singleton.infsep_zero
+
+lemma infsep_pair_eq_inf_to_real (hxy : x ≠ y) :
```

Comment:
```suggestion
lemma infsep_pair_le_to_real_inf (hxy : x ≠ y) :
```
In the order of function applications (without dot notation), the function `to_real` comes first, so it should be first in the lemma name.